schema/default.go

```diff
@@ -0,0 +1,150 @@
+)
+
+const (
+	funcNowName  = "now()"
+	funcUUIDName = "uuid()"
+	funcCUIDName = "cuid()"
+)
+
+func isSupportedDefaultFunction(fnName string) bool {
+	switch fnName {
+	case funcNowName, funcUUIDName, funcCUIDName:
+		return true
+	}
+
+	return !strings.HasSuffix(fnName, "()")
+}
+
+func getDefaultValue(fieldType FieldType) interface{} {
+	switch fieldType {
+	case BoolType:
+		return false
+	case Int32Type, Int64Type, DoubleType:
+		return 0
+	}
+
+	return nil
+}
+
+type FieldDefaulter struct {
+	value     interface{}
+	createdAt bool
+	updatedAt bool
+}
+
+func newDefaulter(createdAt *bool, updatedAt *bool, name string, dataType FieldType, v interface{}) (*FieldDefaulter, error) {
+	if createdAt != nil && *createdAt {
+		return &FieldDefaulter{
+			createdAt: true,
+		}, nil
+	}
+	if updatedAt != nil && *updatedAt {
+		return &FieldDefaulter{
+			updatedAt: true,
+		}, nil
+	}
+
+	defaulter := &FieldDefaulter{}
+	if v == "" {
+		defaulter.value = getDefaultValue(dataType)
+		return defaulter, nil
+	}
+
+	var err error
+	switch ty := v.(type) {
+	case bool:
+		if dataType != BoolType {
+			return nil, errors.InvalidArgument("default value should be true/false only for boolean field '%s'", name)
+		}
+	case json.Number:
+		if dataType == Int32Type || dataType == Int64Type {
+			if _, err = ty.Int64(); err != nil {
+				return nil, errors.InvalidArgument("default value should be numeric for numeric field '%s'", name)
+			}
+		}
+		if dataType == DoubleType {
+			if _, err = ty.Float64(); err != nil {
+				return nil, errors.InvalidArgument("default value should be numeric for numeric field '%s' %s", name, err.Error())
+			}
+		}
+	case string:
+		if dataType != StringType && dataType != DateTimeType && dataType != UUIDType && dataType != ByteType {
+			return nil, errors.InvalidArgument("default value is not supported for '%s' type: '%s'", name, FieldNames[dataType])
+		}
+		if defaulter.createdAt && dataType != DateTimeType {
+			return nil, errors.InvalidArgument("createdAt is only supported for date-time type")
+		}
+		if defaulter.updatedAt && dataType != DateTimeType {
+			return nil, errors.InvalidArgument("updated is only supported for date-time type")
+		}
+		if v.(string) == funcNowName && dataType != DateTimeType {
+			return nil, errors.InvalidArgument("now() is only supported for date-time type")
+		}
+		if !isSupportedDefaultFunction(v.(string)) {
+			return nil, errors.InvalidArgument("'%s' function is not supported", v)
+		}
+	case []any:
+		if dataType != ArrayType {
+			return nil, errors.InvalidArgument("default value is not supported for '%s' type: '%s'", name, FieldNames[dataType])
+		}
+	default:
+		return nil, errors.InvalidArgument("unsupported default value for field '%s' type: '%s'", name, FieldNames[dataType])
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	defaulter.value = v
+
+	return defaulter, nil
+}
+
+func (defaulter *FieldDefaulter) GetValue(createdAt string, updatedAt string) interface{} {
```

Comment:
Do we need two different timestamps here?
I kind of think this should be:

```
func (defaulter *FieldDefaulter) GetValue(timestamp string)

```

And then use that timestamp to set the default value

Reply:
But are you suggesting to not have createdAt/now at all? What will happen in this case?
```
A: @createdAt
B: @default(now())
C: @updatedAt
```

Then during an update, we shouldn't update `A` and `B` as that is set during insert. Similarly, `C` shouldn't be updated during an insert and only during an update. With just timestamp, we may not able to know and still needed this if/else somewhere.

Reply:
Simplified it a bit more to handle createdAt/updatedAt differently i.e. now the caller(the runner) is explicitly using either an insertMutator or updateMutator so that we only either set createdAt or updatedAt.